### DISCUSSION TITLE
Thank you! + Some benchmarks

### DISCUSSION THREAD

**Discussion by andrewdalpino** [zephir-lang/zephir#2189] [General]
Dear Zephir team,

I just wanted to say thank you for all your work. Zephir has played a key role in our development of the [Tensor](https://github.com/RubixML/Tensor) extension which is an optional package that you can install to accelerate machine learning algorithms in [Rubix ML](https://github.com/RubixML/ML) and hopefully more libraries in the future. For us, Zephir really *shines* when the API code is written mostly in Zephir and the important underlying implementations are written in C and then connected using [Custom Optimizers](https://docs.zephir-lang.com/0.12/en/optimizers). We get the ease of use and quick iteration that comes with Zephir and the ability to optimize the code down to bare metal with C.

I also wanted to share with you some recent benchmarks that we are very happy with. By picking off some low hanging fruit using BLAS routines powered by [OpenBLAS](https://en.wikipedia.org/wiki/OpenBLAS) we were able to speed up training on our [MNIST](https://github.com/RubixML/MNIST) example project by 12X - reducing the training time from hours to just minutes.

![tensor-performance-mnist](https://user-images.githubusercontent.com/18690561/108286998-bb14f500-714f-11eb-8f04-fb6ac091124e.png)

![tensor-performance-benchmarks](https://user-images.githubusercontent.com/18690561/108287009-bfd9a900-714f-11eb-9583-82960925be8c.png)

We're hoping to unlock a whole other set of low-hanging fruit once we figure out how to store C data alongside the objects like they describe [here](https://www.zend.com/embedding-c-data-into-php-objects).

So thank you again and keep up the great work!

Andrew

**Comment by rcmcdonald91** [comment]
> We're hoping to unlock a whole other set of low-hanging fruit once we figure out how to store C data alongside the objects like they describe [here](https://www.zend.com/embedding-c-data-into-php-objects).

Any progress on this?
